Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp

```diff
@@ -391,38 +400,84 @@ class Gx2Fitter {
+          ACTS_VERBOSE(
+              "Actor - indices before processing:"
+              << "\n\t"
+              << "result.lastMeasurementIndex: " << result.lastMeasurementIndex
+              << "\n\t"
+              << "result.lastTrackIndex: " << result.lastTrackIndex << "\n\t"
+              << "result.fittedStates->size(): " << result.fittedStates->size())
+
+          // TODO generalize the update of the currentTrackIndex
+          size_t currentTrackIndex = Acts::MultiTrajectoryTraits::kInvalid;
+
+          // Checks if an existing surface is found during the gx2f-iteration.
+          // If not, a new index will be generated afterwards.
+          // During the first iteration, we will always create a new index.
+          if (nUpdate == 0) {
+            ACTS_VERBOSE("   processSurface: nUpdate == 0 decision");
+
+            // Add a full TrackState entry multi trajectory. This allocates
+            // storage for all components, which we will set later.
+            currentTrackIndex =
+                result.fittedStates->addTrackState(mask, result.lastTrackIndex);
+          } else {
+            ACTS_VERBOSE("   processSurface: nUpdate > 0 decision");
+
+            if (result.lastTrackIndex ==
+                Acts::MultiTrajectoryTraits::kInvalid) {
+              currentTrackIndex = 0;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (kInv->0) = "
+                           << currentTrackIndex);
+            } else if (result.lastTrackIndex <
+                       result.fittedStates->size() - 1) {
+              currentTrackIndex = result.lastTrackIndex + 1;
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (n+1) = "
+                           << currentTrackIndex);
+            } else {
+              // Add a full TrackState entry multi trajectory. This allocates
+              // storage for all components, which we will set later.
+              currentTrackIndex = result.fittedStates->addTrackState(
+                  mask, result.lastTrackIndex);
+              ACTS_VERBOSE("   processSurface: currentTrackIndex (ADD NEW)= "
+                           << currentTrackIndex);
+            }
+          }
 
-          // Fill the track state
-          trackStateProxy.predicted() = std::move(boundParams.parameters());
+          auto trackStateProxyRes = detail::getTrackStateProxy(
+              state, stepper, *surface, *result.fittedStates, currentTrackIndex,
+              false, logger(), freeToBoundCorrection, mask, true);
+          if (!trackStateProxyRes.ok()) {
+            result.result = trackStateProxyRes.error();
+          }
+          auto& trackStateProxy = *trackStateProxyRes;
 
           // We have predicted parameters, so calibrate the uncalibrated input
           // measurement
           extensions.calibrator(state.geoContext, *calibrationContext,
                                 sourcelink_it->second, trackStateProxy);
```

Comment:
Do we technically have predicted parameters at this surface? I guess right now the calibrator is ~ a noop, but down the line it might want to get predicted information for actual calibration.

Reply:
Currently, we have them in the trackstate. We can discuss in the feature, if we need them at all. The KF calibrators need them, but maybe we can adapt some things.